### PR TITLE
[LM-11736] Added custom widget support

### PR DESCRIPTION
**Pull request by TyagiLikeMinds**
## Description

**Features**

- Added LMFeedWidgetDataModel in LMFeedCore
- Added LMFeedWidgetContentModel in LMFeedUI
- Fixed widget type from [Widget] to [String:Widget]
- Added meta support in createPost method
- Migrated AddPostRequest and AttachmentMeta to builder pattern
- Added LMFeedBaseCustomCell, LMFeedPostCustomCell in LMFeedUI
- Added support of LMFeedPostCustomCell in LMUIComponents

**Bugs**

- Fixed LMFeedPostHeaderView warning due to background color, replaced bachground color with background view configuration
- Fixed app crash due to wrong initialisation of topic cell in LMFeedCreatePostScreen

**What kind of change does this PR introduce?** (check at least one)

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Hotfix (fix for breaking changes)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Other, please describe:

### Author Checklist

- [x] Assignee added
- [x] Reviewer added
- [x] Description added
- [ ] Unit test cases added

### Reviewer Checklist

- [ ] PR title starts with Jira ticket number prefix
- [ ] If hotfix, Add Hotfix as prefix in PR title
- [ ] Label added
- [ ] Commit messages follows standard format (added, changed, removed, code commented and chore)
- [ ] Description about the Feature/ Bugs added

### CODE REVIEW

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift [hunk 1]
```diff
@@ -48,43 +48,43 @@ final class LMFeedEditPostOperation {
...
-                let attachmentMeta = AttachmentMeta()
-                    .attachmentUrl(medium.url)
-                    .size(medium.size)
-                    .name(medium.name)
-                    .duration(medium.duration ?? 0)
+                var attachmentMeta = AttachmentMeta.Builder()
+                attachmentMeta = attachmentMeta.attachmentUrl(medium.url)
+                attachmentMeta = attachmentMeta.size(medium.size)
+                attachmentMeta = attachmentMeta.name(medium.name)
+                attachmentMeta = attachmentMeta.duration(medium.duration ?? 0)
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift:55]: why we have to do this way?

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift [hunk 2]
```diff
@@ -48,43 +48,43 @@ final class LMFeedEditPostOperation {
...
                 attachments.append(attachment)
             case false:
-                let attachmentMeta = AttachmentMeta()
-                    .attachmentUrl(medium.url)
-                    .size(medium.size)
-                    .name(medium.name)
+                var attachmentMeta = AttachmentMeta.Builder()
+                attachmentMeta = attachmentMeta.attachmentUrl(medium.url)
+                attachmentMeta = attachmentMeta.size(medium.size)
+                attachmentMeta = attachmentMeta.name(medium.name)
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift:67]: why we have to do this way?

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift [hunk 3]
```diff
@@ -48,43 +48,43 @@ final class LMFeedEditPostOperation {
...
-                .size(document.size ?? 0)
-                .pageCount(document.pageCount ?? 0)
-                .name(document.name)
-                .format(document.format ?? ".pdf")
+            var attachmentMeta = AttachmentMeta.Builder()
+            attachmentMeta = attachmentMeta.attachmentUrl(document.url)
+            attachmentMeta = attachmentMeta.size(document.size ?? 0)
+            attachmentMeta = attachmentMeta.pageCount(document.pageCount ?? 0)
+            attachmentMeta = attachmentMeta.name(document.name)
+            attachmentMeta = attachmentMeta .format(document.format ?? ".pdf")
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift:83]: why we have to do this way?

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift [hunk 4]
```diff
@@ -97,31 +97,31 @@ final class LMFeedEditPostOperation {
...
+            var attachmentMeta = AttachmentMeta.Builder()
+            attachmentMeta = attachmentMeta.entityID(poll.id)
+            attachmentMeta = attachmentMeta.title(poll.question)
+            attachmentMeta = attachmentMeta.expiryTime(poll.expiryTime)
+            attachmentMeta = attachmentMeta.pollOptions(poll.options.map({ $0.option }))
+            attachmentMeta = attachmentMeta.multiSelectState(poll.pollSelectType.apiKey)
+            attachmentMeta = attachmentMeta.pollType(poll.isInstantPoll ? "instant" : "deferred")
+            attachmentMeta = attachmentMeta.multSelectNo(poll.pollSelectCount)
+            attachmentMeta = attachmentMeta.isAnonymous(poll.isAnonymous)
+            attachmentMeta = attachmentMeta.allowAddOptions(poll.allowAddOptions)
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift:120]: why we have to do this way?

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift [hunk 5]
```diff
@@ -117,6 +122,24 @@ public struct LMFeedConvertToFeedPost {
...
+    public static func convertToWidget(from data: [LMFeedWidgetDataModel]) -> [LMFeedWidgetContentModel]{
+        data.compactMap{ widget in
+            var lmMeta: LMFeedLMMetaContentModel?
+            if let widgetLMMeta = widget.lmMeta{
+                var pollOptions: [LMFeedPollOptionContentModel]?
+                
+                pollOptions = widgetLMMeta.options.map{
+                    option in
+                    return LMFeedPollOptionContentModel(id: option.id, text: option.text, isSelected: option.isSelected, percentage: option.percentage, uuid: option.uuid, voteCount: option.voteCount)
+                }
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift:134]: shift params to new line

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift [hunk 6]
```diff
@@ -117,6 +122,24 @@ public struct LMFeedConvertToFeedPost {
...
+                
+                lmMeta = LMFeedLMMetaContentModel(options: pollOptions ?? [], pollAnswerText: widgetLMMeta.pollAnswerText, isShowResult: widgetLMMeta.isShowResult, voteCount: widgetLMMeta.voteCount)
+            }
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift:137]: shift params to new line

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift [hunk 7]
```diff
@@ -117,6 +122,24 @@ public struct LMFeedConvertToFeedPost {
...
+            
+            return LMFeedWidgetContentModel.init(id: widget.id, parentEntityID: widget.parentEntityID, parentEntityType: widget.parentEntityType, metadata: widget.metadata, createdAt: widget.createdAt, updatedAt: widget.updatedAt, lmMeta: lmMeta)
+        }
+    }
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift:141]: shift params to new line

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/DataModels/LMFeedPollDataModel.swift [hunk 8]
```diff
@@ -74,11 +74,12 @@ public struct LMFeedPollDataModel {
 }
 
 public extension LMFeedPollDataModel {
-    init?(postID: String, users: [String: User], widgets: [Widget]) {
-        guard let widget = widgets.first(where: { $0.parentEntityID == postID }),
+    init?(postID: String, users: [String: User], widget: Widget?) {
+        guard let widget = widget,
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/DataModels/LMFeedPollDataModel.swift:78]: shift params to new line

**Review by ishaan1607 — CHANGES_REQUESTED** [review]

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/DataModels/LMFeedWidgetDataModel.swift [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import LikeMindsFeed
+
+public struct LMFeedWidgetDataModel {
+    public let id: String?
+    public let parentEntityID: String?
+    public let parentEntityType: String?
+    public let metadata: [String: Any]?
+    public let createdAt: Double?
+    public let updatedAt: Double?
+    public let lmMeta: [String: Any]?
```
[the remark this answers is not in the document]
Comment (TyagiLikeMinds) [reply lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/DataModels/LMFeedWidgetDataModel.swift]: This one is outdated, already been commited in the latest commit

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedCreatePostOperation.swift [hunk 10]
```diff
@@ -43,248 +43,357 @@ final class LMFeedCreatePostOperation {
...
-            .pageCount(pageCount ?? 0)
-            .format("pdf")
-        
+
+        var attachmentMeta = AttachmentMeta.Builder()
+        attachmentMeta = attachmentMeta.attachmentUrl(awsURL)
+        attachmentMeta = attachmentMeta.size(size ?? 0)
+        attachmentMeta = attachmentMeta.name(attachment.fileName)
+        attachmentMeta = attachmentMeta.pageCount(pageCount ?? 0)
+        attachmentMeta = attachmentMeta.format("pdf")
```
[the remark this answers is not in the document]
Comment (TyagiLikeMinds) [reply lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedCreatePostOperation.swift:337]: AttachmentMeta was not on Builder pattern, have to shift it to builder pattern.

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift (hunk 1 above)
Comment (TyagiLikeMinds) [reply lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift:55]: AttachmentMeta was not on Builder pattern, have to shift it to builder pattern.

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift (hunk 2 above)
Comment (TyagiLikeMinds) [reply lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift:67]: AttachmentMeta was not on Builder pattern, have to shift it to builder pattern.

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift (hunk 3 above)
Comment (TyagiLikeMinds) [reply lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift:83]: AttachmentMeta was not on Builder pattern, have to shift it to builder pattern.

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift (hunk 4 above)
Comment (TyagiLikeMinds) [reply lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedEditPostOperation.swift:120]: AttachmentMeta was not on Builder pattern, have to shift it to builder pattern.

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift (hunk 5 above)
Comment (TyagiLikeMinds) [reply lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift:134]: AttachmentMeta was not on Builder pattern, have to shift it to builder pattern.

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift [hunk 11]
```diff
@@ -158,109 +237,126 @@ extension LMFeedConvertToFeedPost {
...
             isShowSubmitButton: isShowSubmitButton,
-            isShowEditVote: !isPollEnded && isPollSubmitted && !pollAttachment.isInstantPoll
+            isShowEditVote: !isPollEnded && isPollSubmitted
+                && !pollAttachment.isInstantPoll
         )
     }
-    
-    public static func isPollSubmitted(options: [LMFeedPollDataModel.Option]) -> Bool {
+
+    public static func isPollSubmitted(options: [LMFeedPollDataModel.Option])
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift:321]: add comments

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift [hunk 12]
```diff
@@ -158,109 +237,126 @@ extension LMFeedConvertToFeedPost {
...
+        -> Bool
+    {
         options.contains(where: { $0.isSelected })
     }
-    
+
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift:326]: add comments

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift [hunk 13]
```diff
@@ -158,109 +237,126 @@ extension LMFeedConvertToFeedPost {
...
         if DateUtility.isEpochTimeInSeconds(time) {
             Int(Date().timeIntervalSince1970) > time
         } else {
             Int(Date().timeIntervalSince1970) > (time / 1000)
         }
     }
-    
-    public static func isMultiChoicePoll(pollSelectCount: Int, pollSelectType: LMFeedPollSelectState) -> Bool {
+
+    public static func isMultiChoicePoll(
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift:335]: add comments

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift [hunk 14]
```diff
@@ -158,109 +237,126 @@ extension LMFeedConvertToFeedPost {
...
-    public static func isShowSubmitButton(isPollEnded: Bool, isMultiChoice: Bool, isPollSubmitted: Bool) -> Bool {
+    public static func isShowSubmitButton(
+        isPollEnded: Bool, isMultiChoice: Bool, isPollSubmitted: Bool
+    ) -> Bool {
         !(isPollEnded || !isMultiChoice || isPollSubmitted)
     }
-    
-    public static func isShowAddOptionButton(isInstantPoll: Bool, isPollSubmitted: Bool, isAllowAddOption: Bool, isPollEnded: Bool, optionCount: Int) -> Bool {
+
+    public static func isShowAddOptionButton(
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Converters/LMFeedConvertToFeedPost.swift:350]: add comments

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedCreatePostOperation.swift [hunk 15]
```diff
@@ -43,248 +43,368 @@ final class LMFeedCreatePostOperation {
...
-        let addPostRequest = AddPostRequest.builder()
-            .heading(heading)
-            .text(content)
-            .attachments(attachments)
-            .addTopics(topics)
-            .build()
-        
-        LMFeedClient.shared.addPost(addPostRequest) { [weak self] response in
+
+    private func createPost(
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedCreatePostOperation.swift:231]: add comments

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedCreatePostOperation.swift [hunk 16]
```diff
@@ -43,248 +43,368 @@ final class LMFeedCreatePostOperation {
...
-                self?.postMessageForCompleteCreatePost(with: response.errorMessage)
+                self?.postMessageForCompleteCreatePost(
+                    with: response.errorMessage)
                 return
             }
             NotificationCenter.default.post(name: .LMPostCreated, object: nil)
         }
     }
-    
+
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedCreatePostOperation.swift:252]: add comments

File: lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedCreatePostOperation.swift [hunk 17]
```diff
@@ -43,248 +43,368 @@ final class LMFeedCreatePostOperation {
...
-        
+
         let attachmentRequest = Attachment()
             .attachmentType(.image)
-            .attachmentMeta(attachmentMeta)
+            .attachmentMeta(attachmentMeta.build())
         return attachmentRequest
     }
-    
+
```
Comment (ishaan1607) [inline lm-feedCore-iOS/lm-feedCore-iOS/Source/Classes/Operations/LMFeedCreatePostOperation.swift:285]: add comments

**Review by ishaan1607 — CHANGES_REQUESTED** [review]

**Review by ishaan1607 — APPROVED** [review]

**Pull request merged**